apps/src/componentLibrary/button/Button.tsx

```diff
@@ -1,36 +1,250 @@
+    throw new Error('Expect href prop when useAsLink is true');
+  }
+
+  if (useAsLink && onClick) {
+    throw new Error(
+      'Expect onClick prop to be undefined when useAsLink is true'
+    );
+  }
+
+  if (!useAsLink && !onClick) {
+    throw new Error('Expect onClick prop when useAsLink is false');
+  }
+
+  if (!useAsLink && href) {
+    throw new Error('Expect href prop to be undefined when useAsLink is false');
+  }
+
+  if (!useAsLink && download) {
+    throw new Error(
+      'Expect download prop to be undefined when useAsLink is false'
+    );
+  }
+
+  if (type !== 'iconOnly' && type !== 'iconBorder' && icon) {
+    throw new Error(
+      'Expect icon prop to be undefined when type is not iconOnly or iconBorder'
+    );
+  }
+};
+
+const spinnerIcon: FontAwesomeV6IconProps = {
+  iconName: 'spinner',
+  iconStyle: 'solid',
+  className: 'fa-spin',
+};
+
+const Button: React.FunctionComponent<ButtonProps> = ({
+  className,
+  id,
+  text,
+  disabled = false,
+  isPending = false,
+  ariaLabel,
+  iconLeft,
+  iconRight,
+  icon,
+  size = 'm',
+  type = 'primary',
+  color = 'purple',
+  buttonType = 'button',
+  /** <a> specific props */
+  useAsLink = false,
+  href,
+  target,
+  download,
+  title,
+  /** <button> specific props */
+  onClick,
+  value,
+  name,
+}) => {
+  const ButtonTag = useAsLink ? 'a' : 'button';
+
+  const tagSpecificProps =
+    ButtonTag === 'a'
+      ? {
+          href: disabled ? undefined : href,
+          target,
+          /** Copied from old button component. Only need it for the older browsers,
+           *  since modern browsers (~2020+ release year secures these vulnerabilities by default) */
+          // Opening links in new tabs with 'target=_blank' is inherently insecure. Unfortunately, we depend
+          // on this functionality in a couple of place. Fortunately, it is possible to partially mitigate some of
+          // the insecurity of this functionality by using the `rel` tag to block some of the potential exploits.
+          // Therefore, we do so here.
+          rel: target === '_blank' ? 'noopener noreferrer' : undefined,
+          download,
+          title,
+        }
+      : {type: buttonType, onClick, value, name};
+
+  // Check if correct props combination is passed
+  useMemo(
+    () =>
+      checkButtonPropsForErrors({
+        type,
+        icon,
+        useAsLink,
+        buttonType,
+        onClick,
+        href,
+        download,
+      }),
+    [type, icon, useAsLink, buttonType, onClick, href, download]
+  );
+
+  /** Handling isPending state content & spinner show logic here.
+     - If there's only text - we show only spinner.
+     - If there's only icon - we show only spinner.
+     - If there's text and iconLeft or both iconLeft and iconRight -> we show spinner on the left + text + iconRight (if it's present).
+     - If there's text and iconRight - we show text + spinner on the right.
+     */
+  const showIcon = icon && !isPending;
```

Comment:
One thought - what do you think about turning this into a function that outputs a list of JSX elements? If i'm understanding this correctly, we'd always have 3 elements at most right? Could we do something like
```
const elements = useMemo(() => {
   const leftElement = isPending ? ...
   const middleElement = ...
   const rightElement = ...
   return [leftElement, middleElement, rightElement];
})
```

Reply:
Correct, good idea, thanks!
Will investigate. Trying to achieve balance between cleanness and readability, not sure if it'll be more readable/understandable based on our logic complexity